File: src/main/java/net/minestom/server/recipe/ShapedRecipe.java
package net.minestom.server.recipe;

import net.minestom.server.item.ItemStack;
import net.minestom.server.network.packet.server.play.DeclareRecipesPacket;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.LinkedList;
import java.util.List;
import java.util.Objects;

public abstract class ShapedRecipe extends Recipe {
    private final int width;
    private final int height;
    private String group;
    private RecipeCategory.Crafting category;
    private final List<DeclareRecipesPacket.Ingredient> ingredients;
    private ItemStack result;
    private boolean showNotification;

    protected ShapedRecipe(@NotNull String recipeId,
                           int width,
                           int height,
                           @NotNull String group,
                            @NotNull RecipeCategory.Crafting category,
                           @Nullable List<DeclareRecipesPacket.Ingredient> ingredients,
                           @NotNull ItemStack result, boolean showNotification) {
        super(Type.SHAPED, recipeId);
        this.width = width;
        this.height = height;
        this.group = group;
        this.category = category;
        this.ingredients = Objects.requireNonNullElseGet(ingredients, LinkedList::new);
        this.result = result;
        this.showNotification = showNotification;
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    @NotNull
    public String getGroup() {
        return group;
    }

    public void setGroup(@NotNull String group) {
        this.group = group;
    }

    @NotNull
    public RecipeCategory.Crafting getCategory() {
        return category;
    }

    public void setCategory(@NotNull RecipeCategory.Crafting category) {
        this.category = category;
    }

    public void addIngredient(DeclareRecipesPacket.Ingredient ingredient) {
        if (ingredients.size() + 1 > width * height) {
            throw new IndexOutOfBoundsException("You cannot add more ingredients than width*height");
        }

        ingredients.add(ingredient);
    }

    @NotNull
    public List<DeclareRecipesPacket.Ingredient> getIngredients() {
        return ingredients;
    }

    @NotNull
    public ItemStack getResult() {
        return result;
    }

    public void setResult(@NotNull ItemStack result) {
        this.result = result;
    }

    public boolean getShowNotification() {
        return showNotification;
    }

    public void setShowNotification(boolean showNotification) {
        this.showNotification = showNotification;
    }
}

File: src/main/java/net/minestom/server/recipe/ShapedRecipe.kt
package net.minestom.server.recipe

import net.minestom.server.network.packet.server.play.DeclareRecipesPacket.Ingredient
import net.minestom.server.item.ItemStack
import net.minestom.server.recipe.Recipe
import java.lang.IndexOutOfBoundsException
import net.minestom.server.network.packet.server.play.DeclareRecipesPacket
import net.minestom.server.network.packet.server.play.DeclareRecipesPacket.DeclaredRecipe
import java.util.concurrent.CopyOnWriteArraySet
import net.minestom.server.recipe.ShapelessRecipe
import net.minestom.server.network.packet.server.play.DeclareRecipesPacket.DeclaredShapelessCraftingRecipe
import net.minestom.server.recipe.ShapedRecipe
import net.minestom.server.network.packet.server.play.DeclareRecipesPacket.DeclaredShapedCraftingRecipe
import net.minestom.server.recipe.SmeltingRecipe
import net.minestom.server.network.packet.server.play.DeclareRecipesPacket.DeclaredSmeltingRecipe
import net.minestom.server.recipe.BlastingRecipe
import net.minestom.server.network.packet.server.play.DeclareRecipesPacket.DeclaredBlastingRecipe
import net.minestom.server.recipe.SmokingRecipe
import net.minestom.server.network.packet.server.play.DeclareRecipesPacket.DeclaredSmokingRecipe
import net.minestom.server.recipe.CampfireCookingRecipe
import net.minestom.server.network.packet.server.play.DeclareRecipesPacket.DeclaredCampfireCookingRecipe
import net.minestom.server.recipe.StonecutterRecipe
import net.minestom.server.network.packet.server.play.DeclareRecipesPacket.DeclaredStonecutterRecipe
import net.minestom.server.recipe.SmithingRecipe
import net.minestom.server.network.packet.server.play.DeclareRecipesPacket.DeclaredSmithingRecipe
import java.util.*
import java.util.function.Supplier

abstract class ShapedRecipe protected constructor(
    recipeId: String,
    val width: Int,
    val height: Int,
    var group: String,
    ingredients: MutableList<Ingredient>?,
    result: ItemStack
) : Recipe(Type.SHAPED, recipeId) {
    private val ingredients: MutableList<Ingredient>
    var result: ItemStack

    init {
        this.ingredients =
            Objects.requireNonNullElseGet(ingredients, Supplier<MutableList<Ingredient>> { LinkedList() })
        this.result = result
    }

    fun addIngredient(ingredient: Ingredient) {
        if (ingredients.size + 1 > width * height) {
            throw IndexOutOfBoundsException("You cannot add more ingredients than width*height")
        }
        ingredients.add(ingredient)
    }

    fun getIngredients(): List<Ingredient> {
        return ingredients
    }
}